R/parallelization.R

```diff
@@ -0,0 +1,101 @@
+#' function is passed in as arguments (a pure function). Any constant arguments
+#' in this case can be supplied through `...` in the usual way.
+#'
+#' However, if the function references an object that is not supplied as an
+#' argument or defined in the function itself, i.e. a free variable, these
+#' should be supplied as a list to `.parallel`. This makes them available to
+#' `.f` in the parallel processes. `purrr` requires these to be explicitly
+#' supplied, rather than try to determine them from analysing `.f`, as this
+#' provides for consistent and reliable behaviour that is hence easy to debug.
+#'
+#' Examples:
+#' \preformatted{
+#' # can just specify TRUE:
+#' map(1:3, stats::runif, .parallel = TRUE)
+#'
+#' # requires 'fun1' to be supplied in a list:
+#' fun1 <- function(x) \{x + x \%\% 2 \}
+#' fun2 <- function(x) \{ x + fun1(x) \}
+#' map(1:3, fun2, .parallel = list(fun1 = fun1))
+#' }
+#'
+#' ## Daemons settings
+#'
+#' How and where parallelization occurs is determined by
+#' [`daemons()`][mirai::daemons]. This is a function from the \pkg{mirai}
+#' package that sets up daemons (persistent background processes that receive
+#' parallel computations) on your local machine or across the network.
+#'
+#' \pkg{purrr} requires these to be set up prior to performing any parallel map
+#' operations. If this is not done, you will see a warning from the \pkg{mirai}
+#' package that it is launching one local daemon - this ensures that the
+#' function proceeds, but as it uses just one daemon, is nearly never what you
+#' would actually want.
+#'
+#' `daemons()`arguments:
+#'
+#' * `n`: the number of daemons to launch on your local machine, e.g.
+#'   `daemons(7)`. As a rule of thumb, for maximum efficiency this should be (at
+#'   most) one less than the number of cores on your machine, leaving one core
+#'   for the main R process.
+#' * `url` and `remote`: used to set up and launch daemons for distributed
+#'   computing over the network. See [daemons][mirai::daemons] function
+#'   documentation for more details.
+#' * None: calling `daemons()` with no arguments returns a summary of the
+#'   current connection status and mirai tasks queued and completed.
+#'
+#' For more details, including on further arguments, see
+#' [daemons][mirai::daemons].
+#'
+#' Resetting daemons:
+#'
+#' `daemons(0)` resets and terminates any existing daemons.
+#'
+#' All daemons automatically terminate when you end your session and there is no
+#' longer a connection. For this reason, it is not required to specifically
+#' terminate daemons in this instance, although it is still good practice to do
+#' so.
+#'
+#' ## with() method
+#'
+#' The `with()` method for daemons provides a convenient way of evaluating a
+#' statement with daemons set up for the duration of the statement.
+#'
+#' As an example, the below is evaluated in parallel using 7 daemons:
+#'
+#' \preformatted{
+#' with(daemons(7), {
```

Comment:
The slightly more standard tidyverse way would be to provide `with_daemons()` and `local_daemons()`. That might be worth doing here. I'm not sure.

Reply:
I thought having it work with `with()` is a nice touch :)

Reply:
I was about to suggest just removing mention of `with()` in the purrr docs for now, but Lionel seems to like it!

It's not at all essential and it's probably more idiomatic just to set daemons at the top of your script for interactive work.

It was designed in the context of Shiny as people tend to not know where to put the `daemons()` call - at the top level or within `server()` etc., so I introduced it so it's easy to do:
```r
app <- shinyApp(server, ui, session)
with(daemons(4), runApp(app))
```

Reply:
It's possible to implement `with_daemons()` and `local_daemons()`, but I prefer not to do it in the initial implementation so as not to overcomplicate things.

I think there's no harm leaving the `with()` method in the docs, but if you think it diverts attention etc. I have no issue with taking the section out.

Reply:
As part of [60dadd1](https://github.com/tidyverse/purrr/pull/1163/commits/60dadd150f27eec120e7a828902582d4f400ce14) I've taken out mention of the `with()` method in the purrr docs as I want it to just focus on 2 things: (i) daemons and (ii) crating.

It's still available of course, and users may find it in the `daemons()` docs.